File: app/backend/routers/holdings.py
"""
/api/holdings - 保有銘柄CRUD
設計ドキュメント準拠
"""
import re
import logging
from fastapi import APIRouter, HTTPException, Query, Depends
from pydantic import BaseModel, field_validator
from typing import Optional, List
import main
from auth import require_auth

logger = logging.getLogger(__name__)

router = APIRouter()

# --- 入力バリデーション ---
_TICKER_RE = re.compile(r"^[A-Z0-9.\-]{1,10}$")
_ACCOUNT_TYPES = {"nisa", "tokutei"}


def _validate_ticker(v: str) -> str:
    v = v.upper()
    if not _TICKER_RE.match(v):
        raise ValueError("Invalid ticker format")
    return v


class HoldingRecord(BaseModel):
    """保有銘柄レコード"""
    id: Optional[str] = None
    user_id: Optional[str] = None
    ticker: str
    shares: float
    avg_price: float
    entry_date: Optional[str] = None
    account_type: Optional[str] = None  # 'nisa', 'tokutei'
    sector: Optional[str] = None
    regime_at_entry: Optional[str] = None  # 'BULL', 'BEAR', 'RECOVERY', 'WEAKENING'
    rs_at_entry: Optional[str] = None  # 'UP', 'FLAT', 'DOWN'
    fx_rate: Optional[float] = 150.0
    target_price: Optional[float] = None
    stop_loss: Optional[float] = None
    thesis: Optional[str] = None
    notes: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class HoldingCreate(BaseModel):
    """保有銘柄作成"""
    ticker: str
    shares: float
    avg_price: float
    entry_date: Optional[str] = None
    account_type: Optional[str] = "tokutei"
    sector: Optional[str] = None
    regime_at_entry: Optional[str] = None
    rs_at_entry: Optional[str] = None
    fx_rate: Optional[float] = 150.0
    target_price: Optional[float] = None
    stop_loss: Optional[float] = None
    thesis: Optional[str] = None
    notes: Optional[str] = None

    @field_validator("ticker")
    @classmethod
    def validate_ticker(cls, v: str) -> str:
        return _validate_ticker(v)

    @field_validator("shares")
    @classmethod
    def validate_shares(cls, v: float) -> float:
        if v <= 0:
            raise ValueError("shares must be positive")
        return v

    @field_validator("avg_price")
    @classmethod
    def validate_price(cls, v: float) -> float:
        if v <= 0:
            raise ValueError("avg_price must be positive")
        return v


class HoldingUpdate(BaseModel):
    """保有銘柄更新"""
    shares: Optional[float] = None
    avg_price: Optional[float] = None
    account_type: Optional[str] = None
    sector: Optional[str] = None
    target_price: Optional[float] = None
    stop_loss: Optional[float] = None
    thesis: Optional[str] = None
    notes: Optional[str] = None


class HoldingsResponse(BaseModel):
    """保有銘柄一覧レスポンス"""
    holdings: List[HoldingRecord]
    total: int
    total_value: Optional[float] = None


@router.get("", response_model=HoldingsResponse)
async def get_holdings(
    user_email: str = Depends(require_auth),
):
    """
    保有銘柄一覧を取得（認証済みユーザーのデータのみ）
    """
    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")

    try:
        query = supabase.table("holdings").select("*").eq("user_id", user_email)

        result = query.order("ticker").execute()

        holdings = [HoldingRecord(**row) for row in result.data]

        # 合計評価額（avg_price * shares）
        total_value = sum(h.shares * h.avg_price for h in holdings)

        return HoldingsResponse(
            holdings=holdings,
            total=len(holdings),
            total_value=total_value
        )

    except Exception as e:
        logger.exception("Holdings API error")
        raise HTTPException(status_code=500, detail=f"Internal server error: {e}")


@router.get("/{ticker}", response_model=HoldingRecord)
async def get_holding(
    ticker: str,
    user_email: str = Depends(require_auth),
):
    """
    特定銘柄の保有情報を取得（認証済みユーザーのデータのみ）
    """
    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")

    ticker = _validate_ticker(ticker)

    try:
        result = (
            supabase.table("holdings")
            .select("*")
            .eq("ticker", ticker)
            .eq("user_id", user_email)
            .limit(1)
            .execute()
        )

        if not result.data:
            raise HTTPException(status_code=404, detail=f"Holding {ticker} not found")

        return HoldingRecord(**result.data[0])

    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Holdings API error")
        raise HTTPException(status_code=500, detail=f"Internal server error: {e}")


@router.post("", response_model=HoldingRecord)
async def create_holding(
    holding: HoldingCreate,
    user_email: str = Depends(require_auth),
):
    """
    保有銘柄を追加（認証済みユーザーに紐付け）
    """
    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")

    try:
        data = {
            "user_id": user_email,
            "ticker": holding.ticker,
            "shares": holding.shares,
            "avg_price": holding.avg_price,
            "entry_date": holding.entry_date,
            "account_type": holding.account_type,
            "sector": holding.sector,
            "regime_at_entry": holding.regime_at_entry,
            "rs_at_entry": holding.rs_at_entry,
            "fx_rate": holding.fx_rate,
            "target_price": holding.target_price,
            "stop_loss": holding.stop_loss,
            "thesis": holding.thesis,
            "notes": holding.notes,
        }

        result = supabase.table("holdings").insert(data).execute()

        if not result.data:
            raise HTTPException(status_code=500, detail="Failed to create holding")

        return HoldingRecord(**result.data[0])

    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Holdings API error")
        raise HTTPException(status_code=500, detail=f"Internal server error: {e}")


@router.put("/{holding_id}", response_model=HoldingRecord)
async def update_holding(
    holding_id: str,
    holding: HoldingUpdate,
    user_email: str = Depends(require_auth),
):
    """
    保有銘柄を更新（所有権検証あり）
    """
    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")

    try:
        # 更新データを構築（Noneでない値のみ）
        update_data = {}
        if holding.shares is not None:
            update_data["shares"] = holding.shares
        if holding.avg_price is not None:
            update_data["avg_price"] = holding.avg_price
        if holding.account_type is not None:
            update_data["account_type"] = holding.account_type
        if holding.sector is not None:
            update_data["sector"] = holding.sector
        if holding.target_price is not None:
            update_data["target_price"] = holding.target_price
        if holding.stop_loss is not None:
            update_data["stop_loss"] = holding.stop_loss
        if holding.thesis is not None:
            update_data["thesis"] = holding.thesis
        if holding.notes is not None:
            update_data["notes"] = holding.notes

        if not update_data:
            raise HTTPException(status_code=400, detail="No update data provided")

        # 所有権を検証しつつ更新
        result = (
            supabase.table("holdings")
            .update(update_data)
            .eq("id", holding_id)
            .eq("user_id", user_email)
            .execute()
        )

        if not result.data:
            raise HTTPException(status_code=404, detail="Holding not found")

        return HoldingRecord(**result.data[0])

    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Holdings API error")
        raise HTTPException(status_code=500, detail=f"Internal server error: {e}")


@router.delete("/{holding_id}")
async def delete_holding(
    holding_id: str,
    user_email: str = Depends(require_auth),
):
    """
    保有銘柄を削除（所有権検証あり）
    """
    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")

    try:
        result = (
            supabase.table("holdings")
            .delete()
            .eq("id", holding_id)
            .eq("user_id", user_email)
            .execute()
        )

        if not result.data:
            raise HTTPException(status_code=404, detail="Holding not found")

        return {"status": "deleted", "holding_id": holding_id}

    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Holdings API error")
        raise HTTPException(status_code=500, detail=f"Internal server error: {e}")


@router.post("/{holding_id}/add-shares")
async def add_shares(
    holding_id: str,
    shares: float = Query(..., gt=0, description="追加株数"),
    price: float = Query(..., gt=0, description="取得単価"),
    user_email: str = Depends(require_auth),
):
    """
    既存の保有銘柄に買い増し（平均取得単価を再計算、所有権検証あり）
    """
    supabase = main.get_supabase()
    if not supabase:
        raise HTTPException(status_code=503, detail="Database not connected")

    try:
        # 所有権を検証しつつ取得
        current = (
            supabase.table("holdings")
            .select("*")
            .eq("id", holding_id)
            .eq("user_id", user_email)
            .single()
            .execute()
        )

        if not current.data:
            raise HTTPException(status_code=404, detail="Holding not found")

        old_shares = current.data["shares"]
        old_price = current.data["avg_price"]

        # 平均取得単価を再計算
        new_shares = old_shares + shares
        new_avg_price = ((old_shares * old_price) + (shares * price)) / new_shares

        # 更新
        result = (
            supabase.table("holdings")
            .update({"shares": new_shares, "avg_price": new_avg_price})
            .eq("id", holding_id)
            .eq("user_id", user_email)
            .execute()
        )

        return {
            "status": "updated",
            "holding_id": holding_id,
            "old_shares": old_shares,
            "new_shares": new_shares,
            "old_avg_price": old_price,
            "new_avg_price": round(new_avg_price, 4),
        }

    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Holdings API error")
        raise HTTPException(status_code=500, detail=f"Internal server error: {e}")

File: app/worker/src/routes/holdings.ts
/**
 * /api/holdings — ポートフォリオ管理
 * ポート元: app/backend/routers/holdings.py (632行)
 */

import type { Env } from '../env';
import { requireAuth, AuthError } from '../middleware/auth';
import { getSupabase } from '../lib/supabase';
import { jsonResponse, errorResponse } from '../lib/response';
import { parseJsonBody, isValidTicker, isPositiveNumber } from '../lib/validation';

export async function handleHoldings(
  request: Request,
  env: Env,
  cors: Record<string, string> | null,
): Promise<Response> {
  const supabase = getSupabase(env);
  const url = new URL(request.url);
  const path = url.pathname.replace('/api/holdings', '') || '/';
  const method = request.method;

  try {
    const userId = await requireAuth(request, env, supabase);

    // Static routes first (before /:param)
    if (path === '/' && method === 'GET') return listHoldings(supabase, userId, cors);
    if (path === '/' && method === 'POST') return createHolding(request, supabase, userId, cors);
    if (path === '/init' && method === 'GET') return initHoldings(supabase, userId, cors);
    if (path === '/portfolio-history' && method === 'GET') return portfolioHistory(url, supabase, userId, cors);
    if (path === '/cash' && method === 'GET') return listCash(supabase, userId, cors);
    if (path === '/cash' && method === 'POST') return createCash(request, supabase, userId, cors);

    // /cash/:id
    const cashMatch = path.match(/^\/cash\/([^/]+)$/);
    if (cashMatch) {
      const cashId = cashMatch[1];
      if (method === 'PUT') return updateCash(request, supabase, userId, cashId, cors);
      if (method === 'DELETE') return deleteCash(supabase, userId, cashId, cors);
    }

    // /:id/add-shares
    const addSharesMatch = path.match(/^\/([^/]+)\/add-shares$/);
    if (addSharesMatch && method === 'POST') {
      return addShares(url, supabase, userId, addSharesMatch[1], cors);
    }

    // /:id (PUT/DELETE) or /:ticker (GET)
    const idMatch = path.match(/^\/([^/]+)$/);
    if (idMatch) {
      const param = idMatch[1];
      if (method === 'GET') return getHolding(supabase, userId, param, cors);
      if (method === 'PUT') return updateHolding(request, supabase, userId, param, cors);
      if (method === 'DELETE') return deleteHolding(supabase, userId, param, cors);
    }

    return errorResponse('Not Found', 404, cors);
  } catch (err) {
    if (err instanceof AuthError) return errorResponse(err.message, err.status, cors);
    return errorResponse('Internal server error', 500, cors);
  }
}

// GET /api/holdings
async function listHoldings(supabase: ReturnType<typeof getSupabase>, userId: string, cors: Record<string, string> | null) {
  const { data } = await supabase
    .table('holdings').select('*').eq('user_id', userId).order('ticker');
  const holdings = data || [];
  const totalValue = holdings.reduce((sum, h) => sum + (h.shares * h.avg_price), 0);
  return jsonResponse({ holdings, total: holdings.length, total_value: totalValue }, 200, cors);
}

// GET /api/holdings/init
async function initHoldings(supabase: ReturnType<typeof getSupabase>, userId: string, cors: Record<string, string> | null) {
  const [holdingsRes, cashRes, fxRes] = await Promise.all([
    supabase.table('holdings').select('*').eq('user_id', userId).order('ticker'),
    supabase.table('cash_balances').select('*').eq('user_id', userId).order('label'),
    supabase.table('market_indicators').select('usdjpy').not('usdjpy', 'is', null).order('date', { ascending: false }).limit(1),
  ]);

  const holdings = holdingsRes.data || [];
  const cash = cashRes.data || [];
  const fxRate = fxRes.data?.[0]?.usdjpy ? Number(fxRes.data[0].usdjpy) : 150.0;
  const totalValue = holdings.reduce((sum, h) => sum + (Number(h.shares) * Number(h.avg_price)), 0);

  return jsonResponse({
    holdings, total: holdings.length, total_value: totalValue,
    cash: { balances: cash, total: cash.length },
    fx_rate: fxRate,
  }, 200, cors);
}

// GET /api/holdings/portfolio-history
async function portfolioHistory(url: URL, supabase: ReturnType<typeof getSupabase>, userId: string, cors: Record<string, string> | null) {
  const months = Math.min(Math.max(Number(url.searchParams.get('months')) || 24, 1), 120);
  const cutoff = new Date(Date.now() - months * 30 * 24 * 60 * 60 * 1000).toISOString().slice(0, 10);

  const { data } = await supabase
    .table('portfolio_snapshots')
    .select('snapshot_date, total_market_value_usd, total_cost_usd, unrealized_pnl_usd, cash_usd, total_assets_usd, holdings_count, fx_rate_usdjpy')
    .eq('user_id', userId)
    .gte('snapshot_date', cutoff)
    .order('snapshot_date', { ascending: true });

  const snapshots = data || [];
  const latest = snapshots.length > 0 ? snapshots[snapshots.length - 1] : null;

  const history = snapshots.map(s => ({
    date: s.snapshot_date,
    total_market_value_usd: Number(s.total_market_value_usd),
    total_cost_usd: Number(s.total_cost_usd),
    unrealized_pnl_usd: Number(s.unrealized_pnl_usd),
    cash_usd: Number(s.cash_usd),
    total_assets_usd: Number(s.total_assets_usd),
    holdings_count: s.holdings_count,
    fx_rate_usdjpy: Number(s.fx_rate_usdjpy || 150.0),
  }));

  return jsonResponse({
    history,
    summary: {
      total_market_value_usd: latest ? Number(latest.total_market_value_usd) : 0,
      total_cost_usd: latest ? Number(latest.total_cost_usd) : 0,
      unrealized_pnl_usd: latest ? Number(latest.unrealized_pnl_usd) : 0,
      total_cash_usd: latest ? Number(latest.cash_usd) : 0,
      total_assets_usd: latest ? Number(latest.total_assets_usd) : 0,
      fx_rate_usdjpy: latest ? Number(latest.fx_rate_usdjpy || 150.0) : 150.0,
    },
  }, 200, cors);
}

// GET /api/holdings/cash
async function listCash(supabase: ReturnType<typeof getSupabase>, userId: string, cors: Record<string, string> | null) {
  const { data } = await supabase.table('cash_balances').select('*').eq('user_id', userId).order('label');
  const balances = data || [];
  return jsonResponse({ balances, total: balances.length }, 200, cors);
}

// POST /api/holdings/cash
async function createCash(request: Request, supabase: ReturnType<typeof getSupabase>, userId: string, cors: Record<string, string> | null) {
  const body = await parseJsonBody(request);
  const label = typeof body.label === 'string' ? body.label.trim() : '';
  if (!label) return errorResponse('label is required', 400, cors);

  const { data, error } = await supabase.table('cash_balances').insert({
    user_id: userId,
    label,
    currency: typeof body.currency === 'string' ? body.currency : 'JPY',
    amount: typeof body.amount === 'number' ? body.amount : 0,
    account_type: typeof body.account_type === 'string' ? body.account_type : null,
  }).select();

  if (error || !data?.length) return errorResponse('Failed to create cash balance', 500, cors);
  return jsonResponse(data[0], 201, cors);
}

// PUT /api/holdings/cash/:id
async function updateCash(request: Request, supabase: ReturnType<typeof getSupabase>, userId: string, cashId: string, cors: Record<string, string> | null) {
  const body = await parseJsonBody(request);
  const updates: Record<string, unknown> = {};
  if (body.label !== undefined) updates.label = body.label;
  if (body.currency !== undefined) updates.currency = body.currency;
  if (body.amount !== undefined) updates.amount = body.amount;
  if (body.account_type !== undefined) updates.account_type = body.account_type;
  if (Object.keys(updates).length === 0) return errorResponse('No update data', 400, cors);

  const { data } = await supabase.table('cash_balances').update(updates).eq('id', cashId).eq('user_id', userId).select();
  if (!data?.length) return errorResponse('Cash balance not found', 404, cors);
  return jsonResponse(data[0], 200, cors);
}

// DELETE /api/holdings/cash/:id
async function deleteCash(supabase: ReturnType<typeof getSupabase>, userId: string, cashId: string, cors: Record<string, string> | null) {
  const { data } = await supabase.table('cash_balances').delete().eq('id', cashId).eq('user_id', userId).select();
  if (!data?.length) return errorResponse('Cash balance not found', 404, cors);
  return jsonResponse({ status: 'deleted', id: cashId }, 200, cors);
}

// GET /api/holdings/:ticker
async function getHolding(supabase: ReturnType<typeof getSupabase>, userId: string, ticker: string, cors: Record<string, string> | null) {
  const t = ticker.toUpperCase();
  if (!isValidTicker(t)) return errorResponse('Invalid ticker format', 400, cors);
  const { data } = await supabase.table('holdings').select('*').eq('ticker', t).eq('user_id', userId).limit(1);
  if (!data?.length) return errorResponse(`Holding ${t} not found`, 404, cors);
  return jsonResponse(data[0], 200, cors);
}

// POST /api/holdings
async function createHolding(request: Request, supabase: ReturnType<typeof getSupabase>, userId: string, cors: Record<string, string> | null) {
  const body = await parseJsonBody(request);
  const ticker = typeof body.ticker === 'string' ? body.ticker.toUpperCase() : '';
  if (!isValidTicker(ticker)) return errorResponse('Invalid ticker format', 400, cors);
  if (!isPositiveNumber(body.shares)) return errorResponse('shares must be positive', 400, cors);
  if (!isPositiveNumber(body.avg_price)) return errorResponse('avg_price must be positive', 400, cors);

  const { data, error } = await supabase.table('holdings').insert({
    user_id: userId,
    ticker,
    shares: body.shares,
    avg_price: body.avg_price,
    entry_date: body.entry_date ?? null,
    account_type: body.account_type ?? 'tokutei',
    sector: body.sector ?? null,
    regime_at_entry: body.regime_at_entry ?? null,
    rs_at_entry: body.rs_at_entry ?? null,
    fx_rate: body.fx_rate ?? 150.0,
    target_price: body.target_price ?? null,
    stop_loss: body.stop_loss ?? null,
    thesis: body.thesis ?? null,
    notes: body.notes ?? null,
  }).select();

  if (error || !data?.length) return errorResponse('Failed to create holding', 500, cors);
  return jsonResponse(data[0], 201, cors);
}

// PUT /api/holdings/:id
async function updateHolding(request: Request, supabase: ReturnType<typeof getSupabase>, userId: string, holdingId: string, cors: Record<string, string> | null) {
  const body = await parseJsonBody(request);
  const updates: Record<string, unknown> = {};
  const fields = ['shares', 'avg_price', 'account_type', 'sector', 'target_price', 'stop_loss', 'thesis', 'notes'];
  for (const f of fields) {
    if (body[f] !== undefined) updates[f] = body[f];
  }
  if ('shares' in updates && !isPositiveNumber(updates.shares)) return errorResponse('shares must be positive', 400, cors);
  if ('avg_price' in updates && !isPositiveNumber(updates.avg_price)) return errorResponse('avg_price must be positive', 400, cors);
  if (Object.keys(updates).length === 0) return errorResponse('No update data provided', 400, cors);

  const { data } = await supabase.table('holdings').update(updates).eq('id', holdingId).eq('user_id', userId).select();
  if (!data?.length) return errorResponse('Holding not found', 404, cors);
  return jsonResponse(data[0], 200, cors);
}

// DELETE /api/holdings/:id
async function deleteHolding(supabase: ReturnType<typeof getSupabase>, userId: string, holdingId: string, cors: Record<string, string> | null) {
  const { data } = await supabase.table('holdings').delete().eq('id', holdingId).eq('user_id', userId).select();
  if (!data?.length) return errorResponse('Holding not found', 404, cors);
  return jsonResponse({ status: 'deleted', holding_id: holdingId }, 200, cors);
}

// POST /api/holdings/:id/add-shares
async function addShares(url: URL, supabase: ReturnType<typeof getSupabase>, userId: string, holdingId: string, cors: Record<string, string> | null) {
  const shares = Number(url.searchParams.get('shares'));
  const price = Number(url.searchParams.get('price'));
  if (!shares || shares <= 0) return errorResponse('shares must be positive', 400, cors);
  if (!price || price <= 0) return errorResponse('price must be positive', 400, cors);

  const { data: current } = await supabase
    .table('holdings').select('*').eq('id', holdingId).eq('user_id', userId).limit(1);
  if (!current?.length) return errorResponse('Holding not found', 404, cors);

  const oldShares = Number(current[0].shares);
  const oldPrice = Number(current[0].avg_price);
  const newShares = oldShares + shares;
  const newAvgPrice = ((oldShares * oldPrice) + (shares * price)) / newShares;

  await supabase.table('holdings')
    .update({ shares: newShares, avg_price: newAvgPrice })
    .eq('id', holdingId).eq('user_id', userId);

  return jsonResponse({
    status: 'updated',
    holding_id: holdingId,
    old_shares: oldShares,
    new_shares: newShares,
    old_avg_price: oldPrice,
    new_avg_price: Math.round(newAvgPrice * 10000) / 10000,
  }, 200, cors);
}
